### PR TITLE
Restore CONTRIBUTING.md from Wiki

### PR DESCRIPTION
**Pull request by MichaelChirico**
Towards #5847.

I first reverted the commit that deleted Contributing.md in favor of the wiki:

00b18cd9

Then I consulted the Wiki history and manually copied over the edits:

https://github.com/Rdatatable/data.table/wiki/Contributing/_history

We should merge this as-is, with no squash commit, to maintain the complete file history.

NB: It might be nice to retain a more faithful copy of the Wiki file history. The current approach loses some metadata: (1) the original commit author and (2) the original commit message.

It was a bit more than I felt like taking on at the moment, but I came across these two seeds that could get you started if you feel so compelled:

 1. The Wiki is apparently "just" a .git repository that you could clone: https://stackoverflow.com/q/15080848/3576984
 2. Then you'd "just" follow normal steps to copy one file between repos while retaining history by applying diffs as patches: https://stackoverflow.com/q/1365541/3576984

### CODE REVIEW

**Comment by jangorecki** [comment]
Thanks, I was going to work on that but struggling to get out of flu for some days already.

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -1,6 +1,130 @@
...
+
+When you file a pull request, you should add some tests to this file with this in mind -- for new features, try to cover possible use cases extensively (we use [Codecov](https://codecov.io/gh/Rdatatable/data.table) to make it a bit easier to see how well you've done to minimally cover any new code you've added); for bug fixes, include a minimal version of the problem you've identified and write a test to ensure that your fix indeed works, and thereby guarantee that your fix continues to work as the codebase is further modified in the future. We encourage you to scroll around in `tests.Rraw` a bit to get a feel for the types of examples that are being created, and how bugs are tested/features evaluated.
```
Comment (jangorecki) [inline .github/CONTRIBUTING.md:52]:
```suggestion
When you file a pull request, you should add some tests to this file with this in mind -- for new features, try to cover possible use cases extensively (we use [Codecov](https://app.codecov.io/gh/Rdatatable/data.table) to make it a bit easier to see how well you've done to minimally cover any new code you've added); for bug fixes, include a minimal version of the problem you've identified and write a test to ensure that your fix indeed works, and thereby guarantee that your fix continues to work as the codebase is further modified in the future. We encourage you to scroll around in `tests.Rraw` a bit to get a feel for the types of examples that are being created, and how bugs are tested/features evaluated.
```

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -1,6 +1,130 @@
...
+8. Ensure that all tests pass by typing `test.data.table()` after installing your branch. It's also better to `R CMD check --as-cran` against your branch source package archive `.tar.gz` file. You may want to add `--no-manual`, `--no-build-vignettes` or `--ignore-vignettes` (R 3.3.0+) options to reduce dependencies required to perform check. PRs with failed tests can't be merged and it is hard to debug every PR and explain why it fails and how to fix it. The lesser the feedback required, the faster it is likely to be merged. Matt has added his dev cycle script [here](https://github.com/Rdatatable/data.table/blob/master/.dev/cc.R) and Pasha has added a Makefile [here](https://github.com/Rdatatable/data.table/blob/master/Makefile). The `.rprofile` setup for how to use Matt's dev cycle script can be found [here](https://github.com/Rdatatable/data.table/issues/5131).
```
Comment (jangorecki) [inline .github/CONTRIBUTING.md:22]:
```suggestion
8. Ensure that all tests pass by typing `test.data.table()` after installing your branch. It's also better to `R CMD check --as-cran` against your branch source package archive `.tar.gz` file. You may want to add `--no-manual`, `--no-build-vignettes` or `--ignore-vignettes` (R 3.3.0+) options to reduce dependencies required to perform check. PRs with failed tests can't be merged and it is hard to debug every PR and explain why it fails and how to fix it. The lesser the feedback required, the faster it is likely to be merged. Matt has added his dev cycle script [here](https://github.com/Rdatatable/data.table/blob/master/.dev/cc.R) and Pasha has added a Makefile [here](https://github.com/Rdatatable/data.table/blob/master/Makefile). The `.Rprofile` setup for how to use Matt's dev cycle script can be found [here](https://github.com/Rdatatable/data.table/issues/5131).
```

File: .github/CONTRIBUTING.md [hunk 3]
```diff
@@ -1,6 +1,130 @@
...
+
+- Use `=` for assignment (not `<-`); see [here](https://github.com/Rdatatable/data.table/pull/3582#discussion_r287075480), [here](https://github.com/Rdatatable/data.table/issues/3590#issuecomment-495776200), [here](https://stackoverflow.com/questions/1741820/what-are-the-differences-between-and-in-r#comment14293879_1742591) and [here](https://twitter.com/kdpsinghlab/status/1044568690346393606)
+- Spacing
+  + 2-space indentation
+  + No trailing white space
+  + In tests use long lines and vertical alignment to easily spot small differences
+  + Argument spacing style: `fun(arg1=value1, arg2=val2, ...)`
+  + Add a whitespace between `if` and opening bracket, also before opening curly bracket: `if (condition) {`
+- Use `L` suffix for integer; i.e. `x[1L]` not `x[1]` (to save coercion)
+- Use `stop(domain=NA, gettextf(fmt, arg1, arg2, ..., domain="R-data.table"))` not `stop(paste(...))` or `stop(sprintf(...))` to facilitate translation as per [Writing R Extensions#1.7 point 2](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Diagnostic-messages)
```
Comment (jangorecki) [inline .github/CONTRIBUTING.md:38]: this is not up to date anymore, we have `stopf`

File: .github/CONTRIBUTING.md [hunk 4]
```diff
@@ -1,6 +1,130 @@
...
+
+`data.table` offers some translations for non-native English speakers (currently, it's limited to just Mandarin). For development, there's a few minor points:
+
+ - You _don't_ need to update the `po` files in your PR. We update these in one PR before each release. This makes it much easier for translators to update the messages in one go, rather than greenlighting each PR.
+ - For R code, when writing error messages (or `warning`s, or `message`s), use a single string where possible. Bad: `stop("This is ", "an error")` vs. Good: `stop("This is an error")`. If you need to construct the error message from several parts, use `gettextf` (and rarely, `ngettext`). Bad: `stop("Error in column ", j, ". Please fix."))` vs. Good: `stop(domain=NA, gettextf("Error in column %d. Please fix.", j, domain="R-data.table"))`. Note the usage of `domain` in both `stop()` and `gettextf()`. More uncommonly, when using `cat()`, always use `gettextf`.
+ - For C code, any `char` array that will be shown to users should be wrapped in `_()` (which marks the array for translation). Bad: `error("Error in column %d. Please fix.", j);` vs. Good: `error(_("Error in column %d. Please fix."), j);`.
```
Comment (jangorecki) [inline .github/CONTRIBUTING.md:46]: to be fair this sounds overly complicated, we can think if we could somehow improve that (not in this PR of course).

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -1,6 +1,130 @@
...
+
+### Testing
+
+`data.table` uses a series of unit tests to exhibit code that is expected to work. These are primarily stored in [`inst/tests/tests.Rraw`](https://github.com/Rdatatable/data.table/blob/master/inst/tests/tests.Rraw). They come primarily from two places -- when new features are implemented, the author constructs minimal examples demonstrating the expected common usage of said feature, including expected failures/invalid use cases (e.g., the [initial assay of `fwrite` includes 28 tests](https://github.com/Rdatatable/data.table/blob/master/inst/tests/tests.Rraw#L9123-L9245)). Second, when kind users such as yourself happen upon some aberrant behavior in their everyday use of `data.table` (typically, some edge case that slipped through the cracks in the coding logic of the original author). We try to be thorough -- for example there were initially [141 tests of `split.data.table`](https://github.com/Rdatatable/data.table/blob/master/inst/tests/tests.Rraw#L8493-L8952), and that number has since grown!
```
Comment (jangorecki) [inline .github/CONTRIBUTING.md:50]: there are two links pointing to particular lines in tests.Rraw, and both are likely to be invalid already because they use master tree, where lines could have been shifted already. We should either replace master with commit hash in the links or remove links.

File: .github/CONTRIBUTING.md [hunk 6]
```diff
@@ -1,6 +1,130 @@
...
+Minimal first time PR
+---------------------
+
+```
+$ cd /tmp      # or anywhere safe to play
+$ git config --global core.autocrlf false   # Windows-only preserve \n in test data
+$ git clone https://github.com/Rdatatable/data.table.git
+$ cd data.table
+$ R CMD build .
+$ R CMD check data.table_1.11.5.tar.gz
```
Comment (jangorecki) [inline .github/CONTRIBUTING.md:75]:
```suggestion
$ R CMD check data.table_1.15.99.tar.gz
```

File: .github/CONTRIBUTING.md [hunk 7]
```diff
@@ -1,6 +1,130 @@
...
+...
+Status: OK
+```
+Congratulations - you've just compiled and tested the very latest version of data.table in development. Everything looks good. Now make your changes. Using an editor of your choice, edit the appropriate `.R`, `.md`, `NEWS` and `tests.Rraw` files. Test your changes : 
+```
+$ R CMD build .
+$ R CMD check data.table_1.11.5.tar.gz
```
Comment (jangorecki) [inline .github/CONTRIBUTING.md:82]:
```suggestion
$ R CMD check data.table_1.15.99.tar.gz
```

File: .github/CONTRIBUTING.md [hunk 8]
```diff
@@ -1,6 +1,130 @@
...
+
+#### Translations
+
+`data.table` offers some translations so that our users can get feedback (errors, warnings, verbose messages) in their native language. Currently we only support Mandarin Chinese.
+
+The data for these translations lives in the `po` folder. You do not need to make any changes here for your PR -- translations are updated in a batch before each CRAN release.
+
+If you are writing R code, please avoid breaking up strings into multiple parts (i.e., wide lines are OK for error messages).
+
+If you are writing C code, please wrap your message strings (`char` arrays) with `_()`, e.g. `error(_("step %d failed"), ii)` instead of `error("step %d failed", ii)`. Common functions for this are `error`, `warning`, and `Rprintf`.
```
Comment (jangorecki) [inline .github/CONTRIBUTING.md:108]: this "translations" sections sounds more friendly than the previous "translations" section. and of course we don't need two.

**Review by jangorecki — COMMENTED** [review]
info about AI written code needs to be somewhere close to the top of the document

**Comment by MichaelChirico** [comment]
Hey Jan, my thinking was to handle this in maybe 3 PRs:

 1. restore wiki to repo with history
 2. update/modernize
 3. finish the new ask about AI-assisted contributions

WDYT?

File: .github/CONTRIBUTING.md (hunk 6 above)
Comment (ben-schwen) [reply .github/CONTRIBUTING.md:75]: Maybe even more future proof?
```suggestion
$ R CMD check data.table_*.tar.gz
```

**Comment by MichaelChirico** [comment]
Merging this, I'll start addressing the comments in the next PR, thanks!

File: .github/CONTRIBUTING.md (hunk 5 above)
Comment (MichaelChirico) [reply .github/CONTRIBUTING.md:50]: Since this is about "how to contribute", probably it's better to point to how tests looked in the accepted PR, e.g. this link for `fwrite()`:

https://github.com/Rdatatable/data.table/pull/1613/files#diff-e3243f3780ce7d303c3317f73945310bfc37e45d193568246246aca20e3270ae

And this for `split()`:

https://github.com/Rdatatable/data.table/commit/5f7a435fea5622bfbe1d5f1ffa99fa94a6a054ae#diff-e3243f3780ce7d303c3317f73945310bfc37e45d193568246246aca20e3270ae

File: .github/CONTRIBUTING.md (hunk 6 above)
Comment (MichaelChirico) [reply .github/CONTRIBUTING.md:75]: This assumes there's no other tar.gz from other version laying around, unless we start with a clean-up step `rm data.table_*.tar.gz`.

File: .github/CONTRIBUTING.md (hunk 8 above)
Comment (MichaelChirico) [reply .github/CONTRIBUTING.md:108]: I think I forgot to delete the old section when I added a re-write to the Wiki. Removed the first one.

File: .github/CONTRIBUTING.md [hunk 9]
```diff
@@ -1,6 +1,130 @@
...
+3. Please create the PR against the `master` branch. You can do that by forking the repository, creating a new branch for your feature/bugfix in the forked project, and then using that as a base for your pull requests. After your first successful merged PR you will very likely be invited to be a project member. This will allow you to create your next branch directly in the project which is easier and more convenient than forking, both for you and for Rdatatable's maintainers. Working on _branches_ on this (Rdatatable) project will _not_ affect the core code, so you can feel free to experiment as a project member; the core code is on the `master` branch, and only Matt & Arun can push/merge code there. Remember to do `git pull upstream your_branch` (where `upstream` is the name of the remote for `Rdatatable/data.table` seen in `git remote -v`) each time you want to add something; this will keep your local branch up to date with remote, in case anyone makes commits you don't yet have locally. This will reduce the number of merge conflicts you will need to deal with. Do not use `git rebase` on a branch where other users are pushing.
```
[the remark this answers is not in the document]
Comment (MichaelChirico) [reply .github/CONTRIBUTING.md:17]: Changed to "data.table Project Member", I couldn't find a link to a current list of who they are, any idea?

**Comment by MichaelChirico** [comment]
Also deleted the Wiki page

File: .github/CONTRIBUTING.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (ben-schwen) [reply .github/CONTRIBUTING.md:17]: [Project member](https://github.com/orgs/Rdatatable/people) since not all visible by the public repo
